groups/bsl/bslstl/bslstl_ratio.t.cpp

```diff
@@ -0,0 +1,349 @@
+        //:   numerator and denominator.
+        //:
+        //: 2 Verify that the comparison operations exist and give 
+        //:   expected results. Check again for the '_v' versions.
+        //:
+        //: 3 Verify that the arithmetic operations exist and give
+        //:   expected results.
+        //
+        // ----------------------------------------------------------------------
+
+        if (verbose) printf("\nTesting ratio operations"
+                            "\n======================================\n");
+
+        if (verbose) printf("\tTest SI unit prefixes.\n");
+        {
+            ASSERT(1 == bsl::deci::num);
+            ASSERT(1 == bsl::centi::num);
+            ASSERT(1 == bsl::milli::num);
+            ASSERT(1 == bsl::micro::num);
+            ASSERT(1 == bsl::nano::num);
+            ASSERT(1 == bsl::pico::num);
+            ASSERT(1 == bsl::femto::num);
+            ASSERT(1 == bsl::atto::num);
+
+            ASSERT(10ULL == bsl::deci::den);
+            ASSERT(100ULL == bsl::centi::den);
+            ASSERT(1000ULL == bsl::milli::den);
+            ASSERT(1000000ULL == bsl::micro::den);
+            ASSERT(1000000000ULL == bsl::nano::den);
+            ASSERT(1000000000000ULL == bsl::pico::den);
+            ASSERT(1000000000000000ULL == bsl::femto::den);
+            ASSERT(1000000000000000000ULL == bsl::atto::den);
+
+            ASSERT(1 == bsl::deca::den);
+            ASSERT(1 == bsl::hecto::den);
+            ASSERT(1 == bsl::kilo::den);
+            ASSERT(1 == bsl::mega::den);
+            ASSERT(1 == bsl::giga::den);
+            ASSERT(1 == bsl::tera::den);
+            ASSERT(1 == bsl::peta::den);
+            ASSERT(1 == bsl::exa::den);
+
+            ASSERT(10ULL == bsl::deca::num);
+            ASSERT(100ULL == bsl::hecto::num);
+            ASSERT(1000ULL == bsl::kilo::num);
+            ASSERT(1000000ULL == bsl::mega::num);
+            ASSERT(1000000000ULL == bsl::giga::num);
+            ASSERT(1000000000000ULL == bsl::tera::num);
+            ASSERT(1000000000000000ULL == bsl::peta::num);
+            ASSERT(1000000000000000000ULL == bsl::exa::num);
+
+        #ifdef BSL_RATIO_SUPPORTS_EXTENDED_SI_TYPEDEFS
+            ASSERT(1 == bsl::zepto::num);
+            ASSERT(1 == bsl::yocto::num);
+            ASSERT(1000000000000000000000ULL == bsl::zepto::den);
+            ASSERT(1000000000000000000000000ULL == bsl::yocto::den);
+
+            ASSERT(1 == std::zetta::den);
+            ASSERT(1 == std::yotta::den);
+            ASSERT(1000000000000000000000ULL == bsl::zetta::num);
+            ASSERT(1000000000000000000000000ULL == bsl::yotta::num);
+        #endif
+        }
+
+        if (verbose) printf("\tTest ratio comparisons.\n");
+        {
+            typedef bsl::ratio<3, 2> three_halfs;
+            typedef bsl::ratio<6, 4> six_fourths;
+            typedef bsl::ratio<0, 1> zero;
+            typedef bsl::ratio<-2, 3> neg_two_thirds;
```

Comment:
Test coverage seems sufficient in this test driver since we're verifying that what we expect to get from the platform's `<ratio>` is all there. However, I would also have one test verifying that the second template parameter of `ratio` does indeed default (to 1).  For completeness, I would also check that `ratio` has the `type` member.

Reply:
Ha! I was about to say "ratio doesn't have a nested type member", and point you to C++reference. But cppreference is wrong about this. So I fixed it there, and will add a test here. Nice catch.

Reply:
And Tim just convinced me that cppreference has it right. I kinda disagree, because you have to scroll down. Whatever.